### PR TITLE
Refactor User#requests and User#incoming_requests

### PR DESCRIPTION
**Pull request by ChrisBr**
to use subselects to improve performance

### CODE REVIEW

File: src/api/app/models/user.rb [hunk 1]
```diff
@@ -768,20 +768,21 @@ def outgoing_requests(search = nil)
...
+      BsRequest.where(id: BsRequestAction.where(target_project_id: projects).select(:bs_request_id))).or(
+        BsRequest.where(id: BsRequestAction.where(target_package_id: packages).select(:bs_request_id))).or(
+          BsRequest.where(id: Review.where(state: :new, user_id: id).select(:bs_request_id)).or(
+            BsRequest.where(id: Review.where(state: :new, project_id: projects).select(:bs_request_id)).or(
+              BsRequest.where(id: Review.where(state: :new, package_id: packages).select(:bs_request_id)).or(
+                BsRequest.where(id: Review.where(state: :new, group_id: groups.pluck(:id)).select(:bs_request_id))
+              )
+            )
+          )
+        ).with_actions_and_reviews
```
Comment (DavidKang) [inline src/api/app/models/user.rb]: This is really needed? :thinking:

Comment (ChrisBr) [inline src/api/app/models/user.rb]: Hm, with actions is at least needed for the view!

Comment (ChrisBr) [inline src/api/app/models/user.rb]: But maybe an includes is even better

File: src/api/app/models/user.rb [hunk 2]
```diff
@@ -768,20 +768,21 @@ def outgoing_requests(search = nil)
...
+    packages = involved_packages.pluck(:id)
+
+    # Test with production data if you change this query to make sure performance does not get worse
+    result = BsRequest.where(creator: login).or(
+      BsRequest.where(id: BsRequestAction.where(target_project_id: projects).select(:bs_request_id))).or(
+        BsRequest.where(id: BsRequestAction.where(target_package_id: packages).select(:bs_request_id))).or(
+          BsRequest.where(id: Review.where(state: :new, user_id: id).select(:bs_request_id)).or(
+            BsRequest.where(id: Review.where(state: :new, project_id: projects).select(:bs_request_id)).or(
+              BsRequest.where(id: Review.where(state: :new, package_id: packages).select(:bs_request_id)).or(
+                BsRequest.where(id: Review.where(state: :new, group_id: groups.pluck(:id)).select(:bs_request_id))
```
Comment (hennevogel) [inline src/api/app/models/user.rb]: Soon you're going to run off the right side of your screen and we have to buy another monitor.... 😂

Comment (ChrisBr) [inline src/api/app/models/user.rb]: Not my fault ....

File: src/api/app/models/bs_request.rb [hunk 3]
```diff
@@ -32,8 +32,6 @@ class SaveError < APIException
   scope :to_accept, -> { where(state: 'new').where('accept_at < ?', DateTime.now) }
   # Scopes for collections
   scope :with_actions, -> { joins(:bs_request_actions).distinct.order(priority: :asc, id: :desc) }
-  scope :with_involved_projects, ->(project_ids) { where(bs_request_actions: { target_project_id: project_ids } ) }
-  scope :with_involved_packages, ->(package_ids) { where(bs_request_actions: { target_package_id: package_ids } ) }
```
Comment (evanrolfe) [inline src/api/app/models/bs_request.rb]: Why get rid of these scopes?

File: src/api/app/models/user.rb [hunk 4]
```diff
@@ -768,20 +768,21 @@ def outgoing_requests(search = nil)
...
-               .where(bs_requests_table[:creator].eq(login)
-                 .or(bs_actions_table[:target_project_id].in(involved_projects.pluck(:id)))
-                 .or(bs_actions_table[:target_package_id].in(involved_packages.pluck(:id)))
-                 .or(reviews_table[:state].eq(:new)
-                     .and(reviews_table[:user_id].eq(id)
-                     .or(reviews_table[:project_id].in(involved_projects.pluck(:id)))
-                     .or(reviews_table[:package_id].in(involved_packages.pluck(:id)))
-                     .or(reviews_table[:group_id].in(groups.pluck(:id))))))
+    projects = involved_projects.pluck(:id)
+    packages = involved_packages.pluck(:id)
```
Comment (evanrolfe) [inline src/api/app/models/user.rb]: Theses variables store ids in them so they should be called `project_ids` and `package_ids`

File: src/api/app/models/user.rb (hunk 2 above)
Comment (evanrolfe) [inline src/api/app/models/user.rb]: Is there a way to make this code more readable? Maybe put these `.where` clauses into scopes?

File: src/api/app/models/user.rb (hunk 4 above)
Comment (ChrisBr) [inline src/api/app/models/user.rb]: You're right!

File: src/api/app/models/bs_request.rb (hunk 3 above)
Comment (ChrisBr) [inline src/api/app/models/bs_request.rb]: Because I don't use them anymore ;)

File: src/api/app/models/user.rb [hunk 5]
```diff
@@ -753,8 +753,8 @@ def declined_requests(search = nil)
 
   # list incoming requests involving this user
   def incoming_requests(search = nil)
-    result = BsRequest.with_involved_projects(involved_projects.pluck(:id)).or(
-      BsRequest.with_involved_packages(involved_packages.pluck(:id))
+    result = BsRequest.where(id: BsRequestAction.where(target_project_id: involved_projects.pluck(:id)).select(:bs_request_id)).or(
+      BsRequest.where(id: BsRequestAction.where(target_package_id: involved_packages.pluck(:id)).select(:bs_request_id))
```
Comment (evanrolfe) [inline src/api/app/models/user.rb]: Can you move this code back into scopes like it was done before? Before is much more readable then now.. (Thats what I meant by my comment in the other file "Why get rid of these scopes?")

File: src/api/app/models/user.rb (hunk 2 above)
Comment (evanrolfe) [inline src/api/app/models/user.rb]: Are you sure that this new query is faster than it was before?

**Comment by adrianschroeter** [comment]
On Mittwoch, 5. Juli 2017, 16:53:02 CEST wrote Evan Rolfe:
> evanrolfe commented on this pull request.
> 
> 
> 
> > -                 .or(reviews_table[:state].eq(:new)
> -                     .and(reviews_table[:user_id].eq(id)
> -                     .or(reviews_table[:project_id].in(involved_projects.pluck(:id)))
> -                     .or(reviews_table[:package_id].in(involved_packages.pluck(:id)))
> -                     .or(reviews_table[:group_id].in(groups.pluck(:id))))))
> +    projects = involved_projects.pluck(:id)
> +    packages = involved_packages.pluck(:id)
> +
> +    # Test with production data if you change this query to make sure performance does not get worse
> +    result = BsRequest.where(creator: login).or(
> +      BsRequest.where(id: BsRequestAction.where(target_project_id: projects).select(:bs_request_id))).or(
> +        BsRequest.where(id: BsRequestAction.where(target_package_id: packages).select(:bs_request_id))).or(
> +          BsRequest.where(id: Review.where(state: :new, user_id: id).select(:bs_request_id)).or(
> +            BsRequest.where(id: Review.where(state: :new, project_id: projects).select(:bs_request_id)).or(
> +              BsRequest.where(id: Review.where(state: :new, package_id: packages).select(:bs_request_id)).or(
> +                BsRequest.where(id: Review.where(state: :new, group_id: groups.pluck(:id)).select(:bs_request_id))
> 
> Are you sure that this new query is faster than it was before?

Could you paste the resulting SQL? It looks a bit like many SELECT calls (sub selects or even single ones?)

Any reason not to use a single SQL call using joins?

-- 

Adrian Schroeter
email: adrian@suse.de

SUSE Linux GmbH, GF: Felix Imendörffer, Jane Smithard, Graham Norton, HRB 21284 (AG Nürnberg) 
 
Maxfeldstraße 5                         
90409 Nürnberg 
Germany

File: src/api/app/models/user.rb (hunk 2 above)
Comment (ChrisBr) [inline src/api/app/models/user.rb]: Yes, A LOT! From > 30 seconds to ~ 1 second for my user on production and I only have ~70 requests.

**Comment by ChrisBr** [comment]
@adrianschroeter Yes, it uses subselects now because it is way faster than doing it in a single SQL. As said from ~30 seconds to ~1 second for my user on production.

Resulting SQL would look like this:

```
SELECT COUNT(*) FROM `bs_requests` WHERE (((`bs_requests`.`creator` = 'my_user' OR `bs_requests`.`id` IN (SELECT `bs_request_actions`.`bs_request_id` FROM `bs_request_actions` WHERE `bs_request_actions`.`target_project_id` IN (42))) OR `bs_requests`.`id` IN (SELECT `bs_request_actions`.`bs_request_id` FROM `bs_request_actions` WHERE 1=0)) OR (`bs_requests`.`id` IN (SELECT `reviews`.`bs_request_id` FROM `reviews` WHERE `reviews`.`state` = 'new' AND `reviews`.`user_id` = 42) OR (`bs_requests`.`id` IN (SELECT `reviews`.`bs_request_id` FROM `reviews` WHERE `reviews`.`state` = 'new' AND `reviews`.`project_id` IN (42)) OR (`bs_requests`.`id` IN (SELECT `reviews`.`bs_request_id` FROM `reviews` WHERE `reviews`.`state` = 'new' AND 1=0) OR `bs_requests`.`id` IN (SELECT `reviews`.`bs_request_id` FROM `reviews` WHERE `reviews`.`state` = 'new' AND 1=0)))))
```

**Comment by ChrisBr** [comment]
Furthermore @evanrolfe, I don't like using the plain Arel methods and would prefer to use ActiveRecord instead.

File: src/api/app/models/user.rb (hunk 2 above)
Comment (hennevogel) [inline src/api/app/models/user.rb]: @evanrolfe hm I find it pretty readable once you've grasped what it's doing...

**Comment by adrianschroeter** [comment]
On Mittwoch, 5. Juli 2017, 17:49:51 CEST wrote Christian Bruckmayer:
> @adrianschroeter Yes, it uses subselects now because it is way faster than doing it in a single SQL. As said from ~30 seconds to ~1 second for my user on production.
> 
> Resulting SQL would look like this:
> 
> ```
> SELECT COUNT(*) FROM `bs_requests` WHERE (((`bs_requests`.`creator` = 'my_user' OR `bs_requests`.`id` IN (SELECT `bs_request_actions`.`bs_request_id` FROM `bs_request_actions` WHERE `bs_request_actions`.`target_project_id` IN (42))) OR `bs_requests`.`id` IN (SELECT `bs_request_actions`.`bs_request_id` FROM `bs_request_actions` WHERE 1=0)) OR (`bs_requests`.`id` IN (SELECT `reviews`.`bs_request_id` FROM `reviews` WHERE `reviews`.`state` = 'new' AND `reviews`.`user_id` = 42) OR (`bs_requests`.`id` IN (SELECT `reviews`.`bs_request_id` FROM `reviews` WHERE `reviews`.`state` = 'new' AND `reviews`.`project_id` IN (42)) OR (`bs_requests`.`id` IN (SELECT `reviews`.`bs_request_id` FROM `reviews` WHERE `reviews`.`state` = 'new' AND 1=0) OR `bs_requests`.`id` IN (SELECT `reviews`.`bs_request_id` FROM `reviews` WHERE `reviews`.`state` = 'new' AND 1=0)))))
> ```

oki, looks good.

-- 

Adrian Schroeter
email: adrian@suse.de

SUSE Linux GmbH, GF: Felix Imendörffer, Jane Smithard, Graham Norton, HRB 21284 (AG Nürnberg) 
 
Maxfeldstraße 5                         
90409 Nürnberg 
Germany

**Comment by ChrisBr** [comment]
@evanrolfe Updated your suggestions and using scopes now.
Should we mention that we use the select because we want to perform subselect for performance reasons?
@DavidKang cc

File: src/api/app/models/user.rb [hunk 6]
```diff
@@ -768,20 +771,26 @@ def outgoing_requests(search = nil)
...
+    package_ids = involved_packages.pluck(:id)
+
+    actions = BsRequestAction.bs_request_ids_of_involved_projects(project_ids).or(
+      BsRequestAction.bs_request_ids_of_involved_packages(package_ids)
+    )
+
+    reviews = Review.bs_request_ids_of_involved_users(id).or(
+      Review.bs_request_ids_of_involved_projects(project_ids).or(
+        Review.bs_request_ids_of_involved_packages(package_ids).or(
+          Review.bs_request_ids_of_involved_groups(groups.pluck(:id))
```
Comment (evanrolfe) [inline src/api/app/models/user.rb]: @ChrisBr thanks, this is much easier to read!

**Comment by evanrolfe** [comment]
> Yes, it uses subselects now because it is way faster than doing it in a single SQL. As said from ~30 seconds to ~1 second for my user on production.

Very nice! LGTM (once the rubocop failulre is fixed)

**Pull request merged**